std/zip.d

```diff
@@ -410,19 +412,26 @@ final class ZipArchive
      *
      * Returns: array representing the entire archive.
      */
-    void[] build()
+    void[] build() @safe pure
```

Comment:
The question here is whether anything that returns void[] can be safe in reality.

Reply:
I don't know, but I think so. As long as operating on `void[]` gets flagged as `@system`, I think that marking this function as `@safe` so long as everything it's doing is `@safe` should be fine. The code that uses the `void[]` then has to worry about being `@trusted` appropriately, but this will have been vetted already, so they won't have to worry about it.

Reply:
Since stdx.allocator was recently changed to no longer return `void[]`, I wonder if it would be better to have a `@safe` overload which returned `ubyte[]` and a `@system` overload which casted the result of the first overload to `void[]`. That way everything is backwards compatible but you side-step some potential holes down the road.

Reply:
> I wonder if it would be better to have a @safe overload which returned ubyte[] and a @system overload which casted the result of the first overload to void[].

The overload wouldn't have to be `@system`, and you wouldn't need a cast. Every array converts implicitly and `@safe`ly to `void[]`.

Reply:
In general, I would argue that `ubyte[]` is better than `void[]` when you're dealing with raw file data and the like, but changing the return value would break code, and even if returning `void[]` were considered `@system` (and I don't think that it is), you can't overload on `@safety` attributes.

Reply:
>In general, I would argue that `ubyte[]` is better than `void[]` when you're dealing with raw file data and the like

I agree 100%.

>changing the return value would break code,

Do you have any examples of code that would break? `ubyte[]` implicitly converts to `void[]` so `void[] whatever = ...` will work fine, and if they were casting it to something else that will work just as well with `ubyte[]` as with `void[]`.

Reply:
The issue is if `auto` is used, because then the variable will end up being typed as `ubyte[]` instead of `void[]`, and any code that expects `void[]` risks breaking. The easiest examples that I can think of the moment would be if the result were passed to a template whose template constraint expected `void[]` and didn't work with `ubyte[]` or if a function were overloaded on `ubyte[]` and `void[]`. Neither is particularly likely, but they're both possible. Given that `ubyte[]` implictly converts to `void[]` and that you can't really operate on `void[]` without casting it, a lot of the typical places where code would break won't, but they're still not the same, so I'd be very nervous about just swapping out the type. I expect that a lot of code would just work, but we can't guarantee that it all will.